spec/controllers/users/totp_setup_controller_spec.rb

```diff
@@ -41,64 +41,59 @@
     context 'when user presents invalid code' do
       before do
         sign_in_as_user
+
+        stub_analytics
+        allow(@analytics).to receive(:track_event)
+
         get :new
         patch :confirm, code: 123
       end
 
-      it 'redirects back to authenticator_setup_path' do
+      it 'redirects with an error message' do
         expect(response).to redirect_to(authenticator_setup_path)
-      end
-
-      it 'sets flash[:error] message' do
         expect(flash[:error]).to eq t('errors.invalid_totp')
-      end
-
-      it 'does not enable TOTP for the current user' do
```

Comment:
`before(:all)` isn't ideal here and is generally not recommended because it can cause unexpected behavior if the user changes state in between tests. In addition, the only thing we can set up in the `before(:all)` block is the user creation. For the test to work in this case, everything else has to be specified in each test, which leads to unnecessary duplication.

I'm not sure what you mean by losing fidelity. What are we losing?